### PR TITLE
adding TimeSystem unit tests

### PR DESCRIPTION
**Pull request by amzn-sean**
Signed-off-by: amzn-sean <75276488+amzn-sean@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzCore/Tests/Time/TimeTests.cpp [hunk 1]
```diff
@@ -77,4 +76,191 @@ namespace UnitTest
...
+        }
+
+        void TearDown() override
+        {
+            m_controlTime = AZ::Time::ZeroTimeUs;
+            m_timeSystem.reset();
+            TeardownAllocator();
+        }
+
+        AZ::TimeUs GetDiff(AZ::TimeUs time1, AZ::TimeUs time2)
```
Comment (hultonha) [inline Code/Framework/AzCore/Tests/Time/TimeTests.cpp:97]: uber nit: This could be `const`

File: Code/Framework/AzCore/Tests/Time/TimeTests.cpp [hunk 2]
```diff
@@ -77,4 +76,191 @@ namespace UnitTest
...
+        const AZ::TimeUs baseline = static_cast<AZ::TimeUs>(AZStd::GetTimeNowMicroSecond()) - m_controlTime;
+        const AZ::TimeUs elapsedTime = m_timeSystem->GetRealElapsedTimeUs();
+
+        const AZ::TimeUs diff = GetDiff(baseline, elapsedTime);
+
+        // elapsedTime should be within 10 microseconds from baseline.
+        EXPECT_LT(diff, AZ::TimeUs{ 10 });
+    }
+
+    TEST_F(TimeSystemTests, TestGetElapsedTimeUs)
```
Comment (hultonha) [inline Code/Framework/AzCore/Tests/Time/TimeTests.cpp:122]: uber nit: Including 'Test' in test names tends to be redundant and could just be deleted

File: Code/Framework/AzCore/Tests/Time/TimeTests.cpp [hunk 3]
```diff
@@ -77,4 +76,191 @@ namespace UnitTest
...
+
+        const AZ::TimeUs diff = GetDiff(baseline, elapsedTime);
+
+        // elapsedTime should be within 10 microseconds from baseline.
+        EXPECT_LT(diff, AZ::TimeUs{ 10 });
+    }
+
+    TEST_F(TimeSystemTests, TestElapsedTimeScales)
+    {
+        // slow down 'time';
```
Comment (hultonha) [inline Code/Framework/AzCore/Tests/Time/TimeTests.cpp:139]: uber nit: redundant ';' 😛

**Review by hultonha — APPROVED** [review]
Tests look great! 😄 Nice work! I just have a few minor nits/suggestions, nothing blocking 👍

File: Code/Framework/AzCore/Tests/Time/TimeTests.cpp [hunk 4]
```diff
@@ -77,4 +76,191 @@ namespace UnitTest
...
+
+        // sleep for a bit to advance time.
+        AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(2));
+
+        // advance the tick delta.
+        const AZ::TimeUs delta = m_timeSystem->AdvanceTickDeltaTimes();
+        const AZ::TimeUs baselineDelta = static_cast<AZ::TimeUs>(AZStd::GetTimeNowMicroSecond()) - baselineStart;
+
+        // the delta should be 1 millisecond
+        EXPECT_EQ(delta, AZ::TimeUs{ 1000 });
```
Comment (moraaar) [inline Code/Framework/AzCore/Tests/Time/TimeTests.cpp:256]: suggestion: To match the override with the expected value, maybe the override value could be put in a variable and use later as the expected delta. Also, use a less trivial number than 1.

````
const auto overrideDelta = AZ::TimeUs{ 3546 };
m_timeSystem->SetSimulationTickDeltaOverride(overrideDelta);
...
EXPECT_EQ(delta, overrideDelta);
````

**Review by moraaar — APPROVED** [review]
Nice tests. Approved. I made a suggestion on one test.

**Review by SergeyAMZN — APPROVED** [review]

File: Code/Framework/AzCore/Tests/Time/TimeTests.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Time/TimeTests.cpp]: Instead of `const` it could be `static`

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**